### ISSUE TITLE
Deambiguated constructor name is not found is some cases

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/ocaml#10011]
Using the 4.12 branch (currently https://github.com/ocaml/ocaml/commit/fead9e4961b3efecbeda4d25243b32fdce4ddf8a), `dns-server.4.6.2` fails to build with the following error message:
```
#=== ERROR while compiling dns-server.4.6.2 ===================================#
# context              2.1.0~beta3 | linux/x86_64 | ocaml-variants.4.12.0+trunk | file:///home/opam/opam-repository
# path                 ~/.opam/4.12.0+trunk/.opam-switch/build/dns-server.4.6.2
# command              ~/.opam/4.12.0+trunk/bin/dune build -p dns-server -j 47
# exit-code            1
# env-file             ~/.opam/log/dns-server-7-994c42.env
# output-file          ~/.opam/log/dns-server-7-994c42.out
### output ###
#     ocamlopt server/.dns_server.objs/native/dns_trie.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12.0+trunk/bin/ocamlopt.opt -w -40 -g -I server/.dns_server.objs/byte -I server/.dns_server.objs/native -I /home/opam/.opam/4.12.0+trunk/lib/astring -I /home/opam/.opam/4.12.0+trunk/lib/bigarray-compat -I /home/opam/.opam/4.12.0+trunk/lib/cstruct -I /home/opam/.opam/4.12.0+trunk/lib/dns -I /home/opam/.opam/4.12.0+trunk/lib/domain-name -I /home/opam/.opam/4.12.0+trunk/lib/duration -I /home/opam/.opam/4.12.0+trunk/lib/fmt -I /home/opam/.opam/4.12.0+trunk/lib/gmap -I /home/opam/.opam/4.12.0+trunk/lib/ipaddr -I /home/opam/.opam/4.12.0+trunk/lib/logs -I /home/opam/.opam/4.12.0+trunk/lib/macaddr -I /home/opam/.opam/4.12.0+trunk/lib/metrics -I /home/opam/.opam/4.12.0+trunk/lib/ptime -I /home/opam/.opam/4.12.0+trunk/lib/randomconv -I /home/opam/.opam/4.12.0+trunk/lib/result -I /home/opam/.opam/4.12.0+trunk/lib/rresult -I /home/opam/.opam/4.12.0+trunk/lib/seq -I /home/opam/.opam/4.12.0+trunk/lib/stdlib-shims -intf-suffix .ml -no-alias-deps -o server/.dns_server.objs/native/dns_trie.cmx -c -impl server/dns_trie.ml)
# File "server/dns_trie.ml", line 270, characters 33-34:
# 270 |             Error (`Empty (name, K Ns))
#                                        ^
# Error: Unbound constructor K
```
This package compiled fine with OCaml 4.11.

The `K` constructor is defined in `Dns.Rr_map` as `type k = K : 'a key -> k`, the `Dns` module is opened at the top in `Dns_trie` and the constructor is found through constructor disambiguation (added in OCaml 4.01)

Here is the code: https://github.com/mirage/ocaml-dns/blob/713138069df05b1b2176c24fc0e6a792d4f59f3b/server/dns_trie.ml#L270

~I've tried to reproduce the failure in the toplevel without success for the past hour, I have no idea what is making this bit of code behave differently but in a nutshell, here is what the code looks like simplified:~
```
module M = struct
  module Sub = struct
    type k = K : int -> k
  end
end

open M

let f = function
  | 1 -> `Error (Sub.K 1)
  | _ -> `Error (K 1)
```
~I've tried adding more ifs and higher-order functions like the original code does but I couldn't find a simple enough test-case that would fail. I don't really know where to look for that but I'm sure there are people here who would know how to get a compelling failing test-case quicker than I could.~

EDIT: compelling test-case, failing on 4.12 but not in < 4.11 bellow

**Comment by kit-ty-kate** [comment]
Nevermind what I said, I just needed to spend 30 more extra minutes to find a test-case:
```
module M = struct
  module Sub = struct
    type t = B : int -> t
    type k = K : t -> k

    let map : (t -> 'a) -> t list -> 'a list = List.map
  end
end

open M

let f =
  Sub.map (function
    | B 1 ->
      if true then
        Error `BreakIt
      else
        Error (`Error (Sub.K (B 1)))
    | B _ as x ->
      if true then
        Error `BreakIt
      else
        Error (`Error (K x))
  ) []
```

**Comment by kit-ty-kate** [comment]
cc @garrigue

**Comment by gasche** [comment]
From a distance this looks like another instance of a change from #9811 (no propagation between branches). In particular, your repro-case was already rejected before, with the same error, if you use `-principal`. If the user program also breaks with `-principal`, this suggests that it should be fixed with explicit module qualification or a type annotation.

**Comment by garrigue** [comment]
This is indeed related to #9811.
Since the type information is not propagated between branches, here we are not able to see that the return type should be 
```
('a, [> `BreakIt | `Error of M.Sub.k ]) result
```
and cannot use this information to resolve constructor ambiguity.
It is pretty easy to fix it (replay not only on unification errors, but also on other errors), but is it really a good idea here?
Is it the only such case you found?

This said, there is something strange here: what is stopped is not the propagation of type information, but marking it as known. This should have no impact on disambiguation. I'm wondering what is the cause.

Also, looking at the result of my fix, there still seems to be something strange occurring, since independently of the principality warning I also get a warning that `x` is unused... A bad interaction between backtracking and usage, it seems.

**Comment by hhugo** [comment]
FYI, I've seen multiple instances of such case while testing the janestreet codebase

**Comment by xavierleroy** [comment]
Is the situation bad enough that we should consider reverting #9811 ?

**Comment by gasche** [comment]
#9811 started as a hard error, and was turned into a warning in *some* situations by adding backtracking logic in the type-checker (ew). Jacques proposed above to expand the backtracking situations, so that the present regression results in a warning as well rather than an error.

#9811 removes a bit of code in the type-checker that (if I get the situation correctly) is painful to have because (1) nobody quite knows what it does precisely and why it is necessary and (2) in some cases it defeats the claimed specification of GADT type inference in OCaml. This causes breakages (we thought that they would only manifest as warnings, but we were wrong), which can always be fixed by adding explicit annotations in the user program.

In the short term, we could decide to revert #9811 to un-break user programs, or that we do want to encourage users to add annotations in these corner cases. In the long term, we probably want #9811 merged for good.

My impression is that right now we don't understand why this user code breaks, this is an unintended impact of #9811. I think it would be reasonable to revert #9811 as long as we don't have a better understanding of what is happening. Once we understand what happens, we may have a better implementation that fixes some of the issues here, or we may be in better position to justify asking more annotations in user programs.

**Comment by hhugo** [comment]
> FYI, I've seen multiple instances of such case while testing the janestreet codebase

As mentioned on another issue, #10006 has solved all instances or the error for me.

**Comment by xavierleroy** [comment]
For the record: #10006 does not solve the error in the example given above (https://github.com/ocaml/ocaml/issues/10011#issuecomment-724327499).

Could we try harder to understand what's going on here?

**Comment by kit-ty-kate** [comment]
Just so we don't overreact to this:

Amongst the packages in opam-repository that I could test, only `dns-server` triggers this error.
If this is an expected side-effect of #9811 it would be ok and trivial to fix this in the package.

**Comment by hhugo** [comment]
> Also, looking at the result of my fix, there still seems to be something strange occurring, since independently of the principality warning I also get a warning that x is unused... A bad interaction between backtracking and usage, it seems.

I've seen two spurious unused variable warnings, while testing with 4.12. Ill try to provide a repro case.

**Comment by garrigue** [comment]
What happens here is clear enough: the strategy of #9811 is to try typing in a stricter mode (without propagation), and, if there was a unification error related to a scope escape, backtrack and type again in a backward compatible mode, emitting a warning.
There was a bug in #9811 that didn't revert the current level upon backtracking, and which caused issues in a number of packages, but this was fixed by #10006.
This new problem is related to the logic of constructor disambiguation: if it could not select a constructor due to lack of information or a unification error, it will revert to the old behavior of looking it in the environment, and fail with an unknown constructor error. At this point there is no way to see that the error comes from a scope escape, so no further attempt is done, and no warning emitted.
There are a few solutions, the simplest one being to backtrack in all cases. However, this exposes to more of the kind of side effects described by @hhugo, so I'm not too enthusiastic. If the goal is to keep 100% compatibility I will do it.

As for the unused variable warnings, I imagine strange thing can happen since we retype things twice.
Are those warnings always accompanied by the ambiguity warning? In that case I would say that it's sort of ok: fixing the ambiguity warning will fix the (spurious) unused variable warnings.
Otherwise this gets more hairy, and we may want to temporarily roll back #9811 (and #10006).

**Comment by trefis** [comment]
> Also, looking at the result of my fix, there still seems to be something strange occurring, since independently of the principality warning I also get a warning that x is unused... A bad interaction between backtracking and usage, it seems.

The first time `mk_cases` is called, `add_pattern_variables` adds a bunch of variables to the environment, and to the global tables used to track usage.
Then, when there is an error and you backtrack, the environment gets backtracked, but not these tables.
You then call `mk_cases` again, which adds a new version of these variables, and end up using this version.
The variables from the first call are then left unused, and reported as such.

This seems reasonably easy to fix / work-around, however I guess (but haven't checked) that the issue would also appear in situations like
```ocaml
| K -> let x = something in (* backtrackable error here *); use x
```
and fixing this in a nice way seems less straightforward.

**Comment by trefis** [comment]
(Of course, an "easy" way out, would be to replace the hash tables by some maps, and snapshot and backtrack those as well)

**Comment by lpw25** [comment]
One thing I can't see in #9811: did someone actually do the opam-wide testing for the affect of making that warning an error? It would great to just remove the new backtracking behaviour and avoid all the various issues it causes.

If we really can't face making the change all in one go, another alternative to backtracking to emit a warning would be a compile-time flag or similar to change between the permissive and restrictive modes for a couple of versions whilst people update their code.

**Comment by Octachron** [comment]
The opam-wide test was done on the 4.140 backport in #9812 .

**Comment by Octachron** [comment]
After some discussions at today's developer meetings, we have decided to revert this changes in 4.12 to get the time to get it completely right in 4.13 .

**Comment by trefis** [comment]
More precisely, I think what's hiding under the "completely right" is this: the initial error was deemed quite hard to understand for users, so it was turned into a warning with a much clearer description of what the issue is. However, to do that, we had to backtrack on bigger scale than we usually do, which is its own source of issues, in this case: spurious warnings that can't be fixed but will magically disappear if the original issue flagged by this new warning is fixed.
This makes the "let's turn the error into a warning" part of these changes counterproductive: if you turn off the warning, you get unfixable warnings.
So for 4.13 the plan is to just have the error, but hopefully with as good of a message than the current warning does.

Until then, we do indeed plan to remove the warning.

**Comment by Octachron** [comment]
Reverted in 4.12 with commit 4d6fcf894a7.

**Comment by kit-ty-kate** [comment]
It looks like this was fixed by https://github.com/ocaml/ocaml/pull/10362 merged in trunk on April 22nd (before 4.12 was branched). This is thus a duplicate of https://github.com/ocaml/ocaml/issues/8575. Closing.
